sql: don't copy ColumnFamilyDescriptor and IndexDescriptor by value in loops

nvb opened the pull request:
`ColumnFamilyDescriptor`s are 80 bytes and `IndexDescriptor`s are 392 bytes, so they're too big to be copying by value when it's trivial to avoid doing so.

I was originally prompted to look into this when I saw the copy in `prepareInsertOrUpdateBatch` account for .59% of a CPU profile.

#### Performance diff (excluding the first commit):
```
name                       old time/op    new time/op    delta
KV/Insert/SQL/rows=1000-4    24.0ms ± 3%    23.6ms ± 2%  -1.88%  (p=0.027 n=10+8)

name                       old alloc/op   new alloc/op   delta
KV/Insert/SQL/rows=1000-4    4.89MB ± 0%    4.89MB ± 0%    ~     (p=0.796 n=10+10)

name                       old allocs/op  new allocs/op  delta
KV/Insert/SQL/rows=1000-4     26.8k ± 0%     26.8k ± 0%    ~     (p=0.838 n=10+10)
```

nvb committed:
sql: don't copy ColumnFamilyDescriptor by value in loops

These descriptors are 80 bytes, so they're a little big to be copying
by value when it's trivial to avoid doing so.

I saw the copy in `prepareInsertOrUpdateBatch` account for .59% of a
CPU profile.

Release note: None

nvb committed:
sql: don't copy IndexDescriptor by value in loops

These descriptors are 392 bytes, so they're far too big to be copying
by value when it's trivial to avoid doing so.

Release note: None

nvb committed:
sql/tests: re-use byte buffer between kvSQL ops

This reduces the number of allocations caused by the benchmark harness.

```
name                    old time/op    new time/op    delta
KV/Delete/SQL/rows=1-4     395µs ± 1%     379µs ± 3%  -4.17%  (p=0.029 n=4+4)

name                    old alloc/op   new alloc/op   delta
KV/Delete/SQL/rows=1-4    35.3kB ± 0%    35.0kB ± 1%  -0.83%  (p=0.057 n=4+4)

name                    old allocs/op  new allocs/op  delta
KV/Delete/SQL/rows=1-4       308 ± 0%       306 ± 0%  -0.81%  (p=0.029 n=4+4)
```

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/37748)
<!-- Reviewable:end -->

nvb committed:
sql: don't accept ColumnDescriptor by value in MarshalColumnValue

These descriptors are 224 bytes, so they're far too big to be copying
by value when it's trivial to avoid doing so.

I saw the following in a profile, which tipped me off to how large
ColumnDescriptors are:
```
     0.01s  0.29%  1.71%      0.04s  1.14%                | github.com/cockroachdb/cockroach/pkg/sql/row.prepareInsertOrUpdateBatch /Users/nathan/Go/src/github.com/cockroachdb/cockroach/pkg/sql/row/writer.go:155
                                             0.01s 25.00% |   runtime.duffcopy /Users/nathan/Go/go/src/runtime/duff_amd64.s:373
                                             0.01s 25.00% |   runtime.duffcopy /Users/nathan/Go/go/src/runtime/duff_amd64.s:385
                                             0.01s 25.00% |   runtime.duffcopy /Users/nathan/Go/go/src/runtime/duff_amd64.s:425
```

Release note: None

jordanlewis reviewed (approved):
> Reviewed 1 of 1 files at r1, 7 of 7 files at r2, 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37748)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz and @jordanlewis)



<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
TFTR!

bors r=jordanlewis

nvb committed:
kv: copy span slice in mergeIntoSpans to avoid "data race"

The gRPC `raceTransport` didn't like that `mergeIntoSpans` was mutating
part of the parallel commit request after it had returned. Copy the
slice in the request before mutating to avoid the perceived data race.

Release note: None

nvanbenschoten commented:
`raceTransport` caught what it considers a "data race" because parallel commits had a code path that would mutate a request after it had returned a response. Fixed in a new commit.

bors r+
